EIPS/eip-2585.md

```diff
@@ -0,0 +1,443 @@
+
+## Test Cases
+
+Tests can be found [here](https://github.com/wighawag/eip-2585/tree/master/contracts/test/)
+
+## Implementation
+
+```solidity
+/* EIP-2585 Minimal Native Meta Transaction Forwarder
+ * This standard defines a universal native meta transaction smart contract
+ * that accept specially crafted Externally Owned Accounts (EOA) signed message
+ * to forward to receiver contract via third parties.
+ *
+ * Written in 2020 by Ronan Sandford
+ *
+ * To the extent possible under law, the author(s) have dedicated all copyright
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * You should have received a copy of the CC0 Public Domain Dedication along
+ * with this software. If not, see
+ * <https://creativecommons.org/publicdomain/zero/1.0/>.
+ *    
+ *       .-''-.  .-./`) .-------.               .`````-.  ,--------.     .-''''-.  ,--------.   
+ *     .'_ _   \ \ .-.')\  _(`)_ \             /   ,-.  \ |   _____|    /  _--.  \ |   _____|   
+ *    / ( ` )   '/ `-' \| (_ o._)|            (___/  |   ||  )          |_( )_ ' | |  )         
+ *   . (_ o _)  | `-'`"`|  (_,_) /_ _    _ _        .'  / |  '----.     (_ o _). / |  '----.    
+ *   |  (_,_)___| .---. |   '-.-'( ' )--( ' )   _.-'_.-'  |_.._ _  '.  .'(_,_).  `.|_.._ _  '.  
+ *   '  \   .---. |   | |   |   (_{;}_)(_{;}_)_/_  .'        ( ' )   \|_( )_    \  |  ( ' )   \ 
+ *    \  `-'    / |   | |   |    (_,_)--(_,_)( ' )(__..--. _(_{;}_)  |(_ o _)   /  |_(_{;}_)  | 
+ *     \       /  |   | /   )               (_{;}_)      ||  (_,_)  /  (_,_)..-' .'|  (_,_)  /  
+ *      `'-..-'   '---' `---'                (_,_)-------' `...__..'     `-....-'   `...__..'   
+ *                                                                                           
+ */
+pragma solidity 0.6.4;
+pragma experimental ABIEncoderV2;
+
+interface ERC1271 {
+    function isValidSignature(bytes calldata data, bytes calldata signature) external view returns (bytes4 magicValue);
+}
+
+interface ERC1654 {
+   function isValidSignature(bytes32 hash, bytes calldata signature) external view returns (bytes4 magicValue);
+}
+
+interface ReplayProtection {
+    function checkAndUpdateNonce(address signer, bytes calldata nonce) external returns (bool);
+}
+
+interface Forwarder {
+
+    enum SignatureType { DIRECT, EIP1654, EIP1271 }
+
+    struct Message {
+        address from;
+        address to;
+        uint256 chainId;
+        address replayProtection;
+        bytes nonce;
+        bytes data;
+        bytes32 innerMessageHash;
+	}
+
+    function forward(
+        Message calldata message,
+        SignatureType signatureType,
+        bytes calldata signature
+    ) external payable;
+}
+
+library SigUtil {
+    function recover(bytes32 hash, bytes memory sig) internal pure returns (address recovered) {
+        require(sig.length == 65, "SIGNATURE_INVALID_LENGTH");
```

Comment:
There is a case for sig.length == 64 (see EIP2098)

relevant code:

	function _recover(bytes32 _hash, bytes memory _sign)
	internal pure returns (address)
	{
		bytes32 r;
		bytes32 s;
		uint8   v;

		if (_sign.length == 65) // 65bytes: (r,s,v) form
		{
			assembly
			{
				r :=         mload(add(_sign, 0x20))
				s :=         mload(add(_sign, 0x40))
				v := byte(0, mload(add(_sign, 0x60)))
			}
		}
		else if (_sign.length == 64) // 64bytes: (r,vs) form → see EIP2098
		{
			assembly
			{
				r :=                mload(add(_sign, 0x20))
				s := and(           mload(add(_sign, 0x40)), 0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
				v := shr(7, byte(0, mload(add(_sign, 0x40))))
			}
		}
		else
		{
			revert('invalid-signature-format');
		}

		if (v < 27) v += 27;
		require(v == 27 || v == 28, 'invalid-signature-v');
		return ecrecover(_hash, v, r, s);
	}